### PR TITLE
[YANG] Enhance the port yang model with new port fields: adv_speeds, interface_type and adv_interface_types

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
**- Why I did it**
Enhance the port yang model with new port fields: adv_speeds, interface_types and adv_interface_types
Refer to HLD https://github.com/Azure/SONiC/pull/732 

**- How I did it**
Add fields adv_speeds, interface_type, adv_interface_types

**- How to verify it**
Add yang unit test and verify build is not broken.

**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: src/sonic-yang-models/tests/yang_model_tests/yangTest.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/yangTest.json:1330]: Is it still 1 and 0 and not on and off?

File: src/sonic-yang-models/tests/yang_model_tests/yangTest.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/yangTest.json:1441]: r they mandatory fields, we can remove those if not needed. We will try to keep the file small.

File: src/sonic-yang-models/tests/yang_model_tests/yangTest.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/yangTest.json:1725]: Same here we can remove these fields everywhere, if they are not mandatory thanks.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-port.yang:78]: it is unclear 1 - 128 means here.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-port.yang:84]: it is unclear 1 - 128 means here.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-port.yang:90]: it is unclear 1 - 128 means here.

File: src/sonic-yang-models/tests/yang_model_tests/yangTest.json
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/tests/yang_model_tests/yangTest.json:1330]: Still 1 and 0.

File: src/sonic-yang-models/tests/yang_model_tests/yangTest.json
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/tests/yang_model_tests/yangTest.json:1441]: I will remove them

**Comment by Junchao-Mellanox** [comment]
Hi @lguohan, @praveen-li, could you please review again? Thanks.

File: src/sonic-yang-models/tests/yang_model_tests/yangTest.json
Comment (lguohan) [reply src/sonic-yang-models/tests/yang_model_tests/yangTest.json]: we have discussed this in the yang community subgroup, the recommendation is to use on|off

**Comment by lguohan** [comment]
@junchao-Mellanox, there are unanswered question, can you address them?

File: src/sonic-yang-models/tests/yang_model_tests/yangTest.json
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/tests/yang_model_tests/yangTest.json]: sure

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [reply src/sonic-yang-models/yang-models/sonic-port.yang]: we should define proper enum here.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [reply src/sonic-yang-models/yang-models/sonic-port.yang]: we should define proper enum here.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [reply src/sonic-yang-models/yang-models/sonic-port.yang]: we should define proper enum here.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/yang-models/sonic-port.yang]: Please check latest code change.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/yang-models/sonic-port.yang]: Please check latest code change.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/yang-models/sonic-port.yang]: Please check latest code change.

**Comment by Junchao-Mellanox** [comment]
Hi @lguohan, @praveen-li, could you please review again? Thanks.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-port.yang:84]: please add description

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-port.yang:74]: please add description

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-port.yang:96]: please add description

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (lguohan) [inline src/sonic-yang-models/yang-models/sonic-port.yang:89]: please add description

**Comment by Junchao-Mellanox** [comment]
Hi @lguohan ， could you please kindly review again?

**Comment by lguohan** [comment]
use the keyword description, please refer to the yang model.

**Comment by lguohan** [comment]
can you resolve the conflict? any update on the comments?

**Comment by Junchao-Mellanox** [comment]
Hi @lguohan, I fixed all comment, could you please help review again? Thanks,

**Comment by liat-grozovik** [comment]
@praveen-li can you please approve? if it is we can go a head and merge.

**Comment by lguohan** [comment]
need to change the title to reflect the pr

**Comment by liat-grozovik** [comment]
> need to change the title to reflect the pr

done. please recheck

**Comment by liat-grozovik** [comment]
@praveen-li and @lguohan any further comments?
We are ready with the PRs of the feature and we would like to have the HLD closed before it.

File: src/sonic-yang-models/tests/yang_model_tests/tests/port.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests/port.json:35]: Should be InvalidValue condition failure

File: src/sonic-yang-models/tests/yang_model_tests/tests/port.json
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests/port.json:54]: Should be InvalidValue condition failure

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-port.yang:79]: So there is no other constraint on adv speed?, it can take any number of values. Kindly point to the correct paragraph in HLD for this plz, I will have a look. Thx

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-port.yang:94]: Kindly point to the correct paragraph in HLD here in comment, thanks again

File: src/sonic-yang-models/yang-models/sonic-types.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-types.yang:118]: Again, if possible, point to the HLD which shows from where these values are coming? thanks

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-port.yang:97]: This allows ["CR4", "all"] as well, if that is unwanted behavior then we should put some condition here.

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-port.yang:79]: This allows [100, 400, "all"] as well, if that is unwanted behavior then we should put some condition here. Also add test for the condition if added.

Comment (Junchao-Mellanox) [reply src/sonic-yang-models/yang-models/sonic-port.yang:79]: Yes, there is no other constraint for adv_speeds, I am following the field "speed" in the same yang model file. HLD: https://github.com/Junchao-Mellanox/SONiC/blob/port-auto-neg/doc/port_auto_neg/port-auto-negotiation-design.md#config-advertised-speeds

File: src/sonic-yang-models/yang-models/sonic-port.yang
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/yang-models/sonic-port.yang:94]: https://github.com/Junchao-Mellanox/SONiC/blob/port-auto-neg/doc/port_auto_neg/port-auto-negotiation-design.md#config-advertised-interface-types

Comment (Junchao-Mellanox) [reply src/sonic-yang-models/yang-models/sonic-port.yang:94]: I cannot add the HLD paragraph to the comment here because HLD is not approved yet. And HLD approval requires YANG model PR approval.

File: src/sonic-yang-models/yang-models/sonic-types.yang
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/yang-models/sonic-types.yang:118]: https://github.com/Junchao-Mellanox/SONiC/blob/port-auto-neg/doc/port_auto_neg/port-auto-negotiation-design.md#config-interface-type

**Comment by Junchao-Mellanox** [comment]
Hi @praveen-li, could you please review again?

**Comment by anshuv-mfst** [comment]
Hi @prsunny - could you please take a look, thanks.

**Comment by keboliu** [comment]
@prsunny  @praveen-li would you please have a look?

**Review by lguohan — APPROVED** [review]
:shipit:

**Comment by lguohan** [comment]
@praveen-li , lgtm, can you also check?

**Comment by anshuv-mfst** [comment]
@praveen-li - could you please review as well.

**Comment by keboliu** [comment]
@praveen-li would you please check?

**Comment by anshuv-mfst** [comment]
Hi @praveen-li - Can you please review and approve at the earliest.

**Pull request merged**
